<font size="8">Homework 2: AirBnB Document Database</font>

Group number: `1`

Group members:
1. `Afonso Ascensão : 20240684`
2. `André Filipe Silva : 20230972`
3. `Joana Esteves : 20240746`
4. `Susana Reis : 20240567`
5. `Tomás Figueiredo : 20240941`

The Homework 2 is comprised of two parts:
1. Data modelling (15 points).
2. Queries to database to answer the questions (25 points).

<font size="6">1. Data Modelling</font>

<font size="4">Congratulations! You’ve been hired as part of the new Data Engineering and Management team in the AirBNB Business Intelligence department. The company is restructuring due to unsatisfactory performance from the previous teams.

Before leaving, the head of the Data Modelling department highlighted several issues:

**Data Storage**: A lot of data about AirBNB listings is stored in a single document. While this approach has some advantages, it has also caused performance issues. Queries are slow, and the team didn’t apply patterns, which could improve performance by optimizing the data model. Indexes were also not used.

**Reviews Growth**: The number of reviews for AirBNB is growing rapidly. Currently, we overwrite reviews regularly, but the Business Intelligence department will benefit from storing all reviews and analyzing them over time.

**Data Errors**: There are errors in the data collection, such as duplicate data entries and incorrect timestamps for transactions. The new team will need to decide how to fix these issues.

**Your Role**: In your new role, you’ll need to consider how each database query is used, how often it is needed, and its impact on reads and writes. You should update the database schema to optimize for business use cases. Use tools like embedding, linking, indexes, and patterns to improve the data model. You may need to create new fields, documents, or collections. Be sure to document the pattern you’re applying and the reasons behind your decisions, especially when dealing with duplication and risks of outdated data.

**Key tasks include**:

1. Streamlining the data collection process.
2. Cleaning up the data and optimizing what will be returned for each use case.
3. Applying the correct patterns to speed up common queries.
4. Ensuring departments get accurate and relevant information from the database.
5. Sharing the updated data model schema with other departments.

**Good Practices**: [Check Chapter 6, Mastering MongoDB]

1. All newly created fields should have capitalized names.
2. New queries should work with the most up-to-date database version. If you make multiple changes, all queries should still work after the final updates.
3. For some queries, you may need to change the database schema.
4. When you are applying specific patterns, like polymorphic, subset, or bucket, name them accordingly. 
5. Document each major transformation using this format:
*“We applied {transformation name} because {reasoning behind it}. We expect {change/result} based on {observable measure, such as query speed, number of documents returned, index use, etc.}.”*

</font>


## **Joana + Tomás**

**Data Cleanup and Schema Adjustments:** [9 points in total] 

1) Before working on the queries below, review the data and adjust the schema based on the typical use case described.

**Typical Use Case**: The most common use of the database is to show property listing information to customers. A query retrieves a listing document from the database. Currently, retrieving a listing takes too long. Decide what information should be included in a typical query and optimize the structure accordingly. For example, customers usually only need a sample of reviews, not all reviews (even though all reviews are stored). They also don’t need past transaction data. Update the document schema to fit this use case. This might involve creating new collections or documents.

**Data Cleanup**: Review the data for any errors (such as transactions that don’t belong to the listing) or unnecessary duplication, and clean it up where needed.

<font size="6">2. Uses for the database</font>

<font size="4">Different AirBNB departments require different analytics from our common database.
Below are the specific questions from various departments: 

**Standard Difficulty Questions:** [2 points per question]

## **Susana**
2)	Once a month, we reward hosts with recognition. Select three superhosts with at least two listings that can accommodate more than four people.

## **Susana**
3)	The company considers inevsting into property to rent. Which bed type is most common in listings with a waterfront and a dishwasher in New York?

## **Susana**
4)	We're considering hiring someone to write reviews professionally. Who wrote the longest review in New York?

## **Susana**
5)	To assess the security of different areas, what is the biggest and smallest (price-security deposit) difference per number of visitors at a property?

## **André**
6)  Identify areas by whether they are typically used for short breaks, like weekend mini breaks, or whether they are more suitable for long trips. This information support targeted advertising of different customer types. It is not expected to change much over time so we won’t look to update it, we just require current view. What is the average duration of stay (in nights) per type of property per city (you can use the maximum_nights to measure length of stays)? For each property type return the city with the highest and lowest average value.

**Advanced Difficulty Questions (Consider database optimization for these queries):** [3 points per question]

## **Afonso**
7)	We are creating a new webpage for hosts when setting up their account. It will list suggested typical amenities. This data will need to be available every time a host registers a property but is not expected to change very much. The starting point for the list will be all unique amenities currently listed in properties (across all documents). Optimise the database for this use case and show how the data should be queried.

## **Afonso**
8)	We plan to rtack our reviewers better. We want to create a webpage that shows the top 20 reviewers and the count of the number of reviews of each of these reviewers. This webpage should be kept up to date. It should also have a link to return the number of reviews for a given reviewer ID or Name (show how to query for number of reviews by ID or query quickly).

## **André**
9)	For each property we store review scores across different metrics (accuracy, check-in, cleanliness etc). We consider adding more metrics, although there is no clarity on what these will be. We want to be able to easily query the average score across all of these metrics, including any new metrics that might be added without changing the query. Adjust the data model so this can be done and show the query for an example property.

## **André**
10)	We aim to have better access to information about transaction, we wish to develop a search engine that can calculate the average value of transactions in a given period of time quickly for a given property.

## **André**
11)	We wish to have a summary webpage that displays information about our top destinations. This webpage should display for each of the top 10 cities some basic information about our operations in the area (number of properties by type for example, average price by type) but you can choose the metrics. For each of the top 10 cities it should also provide some basic information about the top 3 properties in each city (price, number of review, whatever you think useful) to show an example of the properties available in the area. We would like to keep this webpage up to date as information changes.

**Database updates:** [2 points per question]
## **Joana + Tomás** 
After optimizing the database, show how to complete the following updates. You can create fictional data. Ensure that previous data does not become stale:

12) Add a new property with a new host in one of the top 10 cities. The host selects the top 10 most common amenities to list.

13) Add a new review from one of our top 20 reviewers for this new property.

14) Add a new review metric called 'x_factor' with a score of 10. Show that the average score across all metrics is correctly calculated for this listing, using the previously developed query.


# Setup

1. run Jupyter 
2. run Docker
3. create MongoDB container in Command Prompt: make sure you don't have 'leftover' container from previous runs

        docker run --name mongodb -d -e MONGO_INITDB_ROOT_USERNAME=AzureDiamond -e MONGO_INITDB_ROOT_PASSWORD=hunter2 -p 27017:27017 mongo

4. run Studio3T

5. create New Connection OR reconnect to existing Connection:

    5.A	In Studio3T create Connection:
            a.	Press Connection -> New Connection
            b.	Insert credentials into URI field: mongodb://AzureDiamond:hunter2@localhost:27017
            c.	Press Test Connection
            d.	Assign a name to the Connection (top empty line) -> Save
            e.	Press Connect
            
    5.B In Studio3T reconnect to Connection:
            a. press Connect (top left corner)
            b. choose MongoDB connection  -> press Connect (MongoDB container must be runnning by this time)
            
6.	In Studio3T import the Database:
            a.	Press Import
            b.	Choose BSON mongodump archive 
            c.	Find your database path
            d.	Select All file formats and choose the database file ‘sampledata.archive’
            e.	Press Run: be patient while the database loads
            f.	Inspect the collections and documents

Done!

[PyMongo documentation](https://pymongo.readthedocs.io/en/stable/api/pymongo/collection.html)

In [1]:
# Python Connector

!pip install pymongo
# or #!conda install -y pymongo

from datetime import datetime
from pprint import pprint
import time
from bson.objectid import ObjectId

from pymongo import MongoClient

user="AzureDiamond"
password="hunter2"
host="localhost"
port="27017"
protocol="mongodb"

client = MongoClient(f"{protocol}://{user}:{password}@{host}:{port}")

# Database check
db = client.sample_airbnb 
print(f"Database info: {db}\n")
db.name 

Database info: Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'sample_airbnb')



'sample_airbnb'

# 1. Data Cleanup and Schema Adjustments

## Loading and Inspecting database


In [2]:
# Collections are inside our Database 'sample_analytics'

collection_list = db.list_collection_names()
print(f"The database contains {len(collection_list)} collections")
print(f"All collections: {collection_list[0:]}")
print(f"Collection {collection_list[0]} contains {db[collection_list[0]].count_documents({})} documents")

The database contains 1 collections
All collections: ['listingsAndReviews_new']
Collection listingsAndReviews_new contains 5555 documents


In [3]:
# Get one sample
pprint(db.listingsAndReviews_new.find_one())

{'_id': '10006546',
 'access': 'We are always available to help guests. The house is fully '
           'available to guests. We are always ready to assist guests. when '
           'possible we pick the guests at the airport.  This service transfer '
           'have a cost per person. We will also have service "meal at home" '
           'with a diverse menu and the taste of each. Enjoy the moment!',
 'accommodates': 8,
 'address': {'country': 'Portugal',
             'country_code': 'PT',
             'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, '
                                'Vitória',
             'location': {'coordinates': [-8.61308, 41.1413],
                          'is_location_exact': False,
                          'type': 'Point'},
             'market': 'Porto',
             'street': 'Porto, Porto, Portugal',
             'suburb': ''},
 'amenities': ['TV',
               'Cable TV',
               'Wifi',
               'Kitchen',
              

In [4]:
# Get unique fields from the documents 

unique_fields = set()

for document in db.listingsAndReviews_new.find():
    for field in document.keys():
        unique_fields.add(field)

for field in sorted(unique_fields):
    print(field)

_id
access
accommodates
address
amenities
availability
bathrooms
bed_type
bedrooms
beds
calendar_last_scraped
cancellation_policy
cleaning_fee
description
extra_people
first_review
guests_included
host_about
host_has_profile_pic
host_id
host_identity_verified
host_is_superhost
host_listings_count
host_location
host_name
host_neighbourhood
host_picture_url
host_response_rate
host_response_time
host_thumbnail_url
host_total_listings_count
host_url
host_verifications
house_rules
images
interaction
last_review
last_scraped
listing_url
maximum_nights
minimum_nights
monthly_price
name
neighborhood_overview
notes
number_of_reviews
price
property_type
review_scores_checkin
review_scores_cleanliness
review_scores_communication
review_scores_location
review_scores_rating
review_scores_value
reviews
reviews_copy1
reviews_copy2
reviews_copy3
reviews_copy4
reviews_per_month
room_type
security_deposit
space
summary
transactions
transit
weekly_price


Structure of the document:
- The document is structured in a way where information about images, adress, availability, reviews and transactions is nested inside the listings' documents.
- The host information should either be nested within a dedicated host field or stored in a separate collection.
- We will analyze the possibility of creating other collections:
    - Images, adress and availability nested fields hold information that is relevant when a user wants to check a listing and the amount of information is small and unique to each listing. As such, it makes sense to keep them as nested fields to avoid queries across collections.
    - New collections to consider: Hosts, Reviews and Transactions. 

## Collections

### Hosts:

Check nº of hosts, max and average listings per host - to understand degree of duplication:

In [21]:
# Total nº of hosts 

pipeline = list( 
    db.listingsAndReviews_new.aggregate([ 
        {"$group": {"_id": "$host_id"}}, 
        {"$count": "total_hosts"},
        {"$project": {"_id": False}} 
        ] ) ) 

pprint(pipeline)

[{'total_hosts': 5104}]


In [22]:
# Total number of listings per host - top 10

pipeline = list( 
    db.listingsAndReviews_new.aggregate([ 
        {"$group": {"_id": "$host_id", "nr_listings": {"$max": "$host_total_listings_count"}}}, 
        {"$sort": {"nr_listings": -1}}, 
        {"$limit": 10},
        {"$project": {"_id": False}} 
        ] ) ) 

pprint(pipeline)

[{'nr_listings': 1198},
 {'nr_listings': 799},
 {'nr_listings': 508},
 {'nr_listings': 472},
 {'nr_listings': 429},
 {'nr_listings': 372},
 {'nr_listings': 324},
 {'nr_listings': 320},
 {'nr_listings': 294},
 {'nr_listings': 294}]


In [23]:
# Average number of listings per host

pipeline = list( 
    db.listingsAndReviews_new.aggregate([ 
        {"$group": {"_id": None, "avg_listings": {"$avg": "$host_total_listings_count"}}},
        {"$project": {"_id": False}} 
        ] ) ) 

pprint(pipeline)

[{'avg_listings': 14.405760576057606}]


**Decision**: Keep host related information nested inside the listing document. Points considered and justification:
- **Host related operations vs. listing related operations**: Usually users search by location, price, amenities, so the main logic revolves around listing. The interest for a specific host information comes as secondary from the interest in the listing, meaning that most of the time we will retrieve the listing and need complementary information on the host, rather than retrieve the host and then look at the listings he has. A separate host document would only be beneficial if querying all listings for a specific host is common, since we have more listings related operations, having the host informations inside the listing document ensures no need for additional queries across different collections.

- **Frequency of updating host fields**: Having separate host documents in another collection makes updating fields easier, because we only have to update one document, instead of updating all listing documents where that host is. However, most fields are unlikely to be updated frequently, host information often remains the same,  so keeping them inside the listing document prevents unnecessary complexity. 

- **Duplicated information**: Separate host document prevents duplication, however one listing only has one host and keeping the host information inside the listing does not cause the document to be very long, as such the duplication may be justified considering the previous points. On average each host has 15 listings so the duplication is not very extreme, but we recognize there are some cases where the duplication will be heavy (e.g. max value of 1198).

### Reviews:

In [35]:
pipeline = list(
    db.listingsAndReviews_new.aggregate([
        {"$project": {"reviews": 1 }},
        {"$limit": 1}
    ]))

pprint(pipeline)

[{'_id': '10006546',
  'reviews': [{'_id': '58663741',
               'comments': 'A casa da Ana e do Gonçalo foram o local escolhido '
                           'para a passagem de ano com um grupo de amigos. '
                           'Fomos super bem recebidos com uma grande simpatia '
                           'e predisposição a ajudar com qualquer coisa que '
                           'fosse necessário.\r\n'
                           'A casa era ainda melhor do que parecia nas fotos, '
                           'totalmente equipada, com mantas, aquecedor e tudo '
                           'o que pudessemos precisar.\r\n'
                           'A localização não podia ser melhor! Não há melhor '
                           'do que acordar de manhã e ao virar da esquina '
                           'estar a ribeira do Porto.',
               'date': datetime.datetime(2016, 1, 3, 5, 0),
               'listing_id': '10006546',
               'reviewer_id': '51483096',
   

Check max and average reviews per listing:

In [ ]:
# Total number of reviews per listing - top 10

pipeline = list( 
    db.listingsAndReviews_new.aggregate([ 
        {"$project": {"listing_id": 1, "reviews_count": {"$size": "$reviews"}}},
        {"$sort": {"reviews_count": -1}}, 
        {"$limit": 10},
        {"$project": {"_id": False}} 
        ] ) ) 

pprint(pipeline)

[{'reviews_count': 533},
 {'reviews_count': 469},
 {'reviews_count': 463},
 {'reviews_count': 420},
 {'reviews_count': 408},
 {'reviews_count': 402},
 {'reviews_count': 399},
 {'reviews_count': 397},
 {'reviews_count': 391},
 {'reviews_count': 385}]


In [ ]:
# Average number of reviews per listing

pipeline = list( 
    db.listingsAndReviews_new.aggregate([ 
         {"$project": {"listing_id": 1, "reviews_count": {"$size": "$reviews"}}},
        {"$group": {"_id": None,"avg_reviews": {"$avg": "$reviews_count"}}},
        {"$project": {"_id": False}} 
        ] ) ) 

pprint(pipeline)

[{'avg_reviews': 26.965256525652567}]


**Decision**: Separate reviews and listings in two collections, so we can have review specific documents. Points considered and justification: 
- **Size of the document**: The listings have on average 27 reviews, with some cases having hundreds of reviews nested in a single listing. The field “comments” often holds values that are long lines of text. Additionally, with time this problem only tends to get worse as new reviews are added and larger documents risk hitting MongoDB’s document size limit. Given this scenario it would be beneficial to have a new collection for review documents.
- **Review related operations vs. listing related operations**: Usually a user will be interested in a specific listing and want to look at reviews related to that listing. This means that a separate reviews document will lead to additional queries across collections in some cases. However, it is also common to have review related queries (e.g. average review score per listing), and if reviews have a separate collection we will only need to load review related information, avoiding loading all the remaining information that is in the listing’s documents.
- **Keep some info in the listing's documents**: Aditionally, we will keep 3 reviews nested inside the listings' documents for efficient querying.
- **Implications**: We need to ensure consistency between collections. For example, when a new review is added we need to update some fields in the listings' documents, like: last_review and number_of_reviews.

### Transactions:

In [24]:
pipeline = list(
    db.listingsAndReviews_new.aggregate([
        {"$project": {"transactions": 1 }},
        {"$limit": 1}
    ]))

pprint(pipeline)

[{'_id': '10006546',
  'transactions': {'bucket_end_date': datetime.datetime(2016, 11, 23, 0, 0),
                   'bucket_start_date': datetime.datetime(1979, 10, 4, 0, 0),
                   'transaction_count': 21,
                   'transactions': [{'date': datetime.datetime(2008, 8, 12, 0, 0),
                                     'price': '132.106378168429131392258568666875362396240234375'},
                                    {'date': datetime.datetime(1999, 3, 9, 0, 0),
                                     'price': '23.70647577039597564407813479192554950714111328125'},
                                    {'date': datetime.datetime(2016, 6, 15, 0, 0),
                                     'price': '6.88059855597928216042191706947050988674163818359375'},
                                    {'date': datetime.datetime(2015, 12, 17, 0, 0),
                                     'price': '2.585702399312638899431249228655360639095306396484375'},
                                    {'da

In [29]:
# Average number of transactions per listing

pipeline = list( 
    db.listingsAndReviews_new.aggregate([ 
        {"$project": {"transactions": 1}},
        {"$group": {"_id": None, "avg_transactions": {"$avg": "$transactions.transaction_count"}}},
        {"$project": {"_id": False}} 
        ] ) ) 

pprint(pipeline)

[{'avg_transactions': 56.0023402340234}]


**Decision**: Separate transactions and listings into two collections, so that transaction-specific data can be managed independently. Points considered and justification:

- **Size of the document**: Currently we have on average 56 transactions nested inside the listing's documents and transaction data tends to grow over time as more transactions occur for each listing. This results in larger and slower to query documents, with a risk of hitting MongoDB’s document size limit. By moving transactions into a separate collection, we can keep the listing documents smaller and more manageable.

- **Listing-related operations vs. transaction-related operations**: The primary use case is to display property listings to customers, who generally do not need past transaction data. Including transaction data in the listings collection could slow down queries focused on listings. By having a separate transactions collection, listing-related queries can remain fast and efficient and transaction-related queries can be performed independently on the transactions collection.

## Duplicates

### Listings:

In [30]:
# Check listings with duplicated ids 

pipeline = list(
    db.listingsAndReviews_new.aggregate([
        {"$group": { "_id": "$_id", "count": { "$sum": 1 }} },
        { "$match": {"count": { "$gt": 1 }}}
    ])
)

pprint(pipeline)

[]


In [31]:
# check listings w/ duplicated urls
pipeline = list(
    db.listingsAndReviews_new.aggregate([
        {"$group": { "_id": "$listing_url", "count": { "$sum": 1 }} },
        { "$match": {"count": { "$gt": 1 }}}
    ])
)

pprint(pipeline)

[]


In [9]:
# Check listings with duplicates names 

pipeline = list(
    db.listingsAndReviews_new.aggregate([
        {"$group": { "_id": "$name", "count": { "$sum": 1 }} },
        { "$match": {"count": { "$gt": 1 }}}
    ])
)

pprint(pipeline)

[{'_id': 'Merkezi Konum', 'count': 2},
 {'_id': 'Feel like home', 'count': 2},
 {'_id': 'Great Location In Wakiki, Walking Distance To Beach, Shopping, '
         'Dining!',
  'count': 2},
 {'_id': 'Quarto moradia luxo', 'count': 2},
 {'_id': '', 'count': 8},
 {'_id': 'İstanbul Birden fazla bölümden oluşan bina', 'count': 4},
 {'_id': 'Elegant Boavista', 'count': 2},
 {'_id': 'Triple room with private bathroom', 'count': 2},
 {'_id': 'Cozy apartment', 'count': 2}]


In [10]:
# Check listings with duplicated descriptions

pipeline = list(
    db.listingsAndReviews_new.aggregate([
        {"$group": { "_id": "$description", "count": { "$sum": 1 }} },
        { "$match": {"count": { "$gt": 1 }}}
    ])
)

pprint(pipeline)

[{'_id': 'Maison avec 3 chambres à V.N.Gaia, quartier calme, à proximité des '
         'transports (bus\\train), à 15 minutes du quartier historique de '
         'Porto et à 3 km de la plage. Chambres confortables, idéales pour les '
         'gens propres, polis et respectueux. Les clients ont accès à la '
         'plupart de la maison. La cuisine est exclusive aux invités, partagée '
         "seulement s'il y a d'autres invités dans la même période. TAXE "
         'TOURISTIQUE: Été: 2€ / personne / nuit (avril-septembre) Hiver: 1€ / '
         'personne / nuit (octobre-mars) Agréable. La maison est assez grande '
         'pour offrir une bonne privacité. Les clients ont accès à la plupart '
         "de la maison. La cuisine est exclusive, seulement partagée s'il y a "
         "d'autres invités dans la même période. Je vis ici avec mes parents "
         'et nous sommes toujours disponibles pour aider. Je suis également '
         'accessible par téléphone portable. Lieu qui é

In [12]:
# Check listings with duplicated name and description

pipeline = list(
    db.listingsAndReviews_new.aggregate([
        {"$group": { "_id": {"description": "$description", "name": "$name"}, 
                    "count": { "$sum": 1 },
                    "documents": { "$push": "$_id" }} },

        { "$match": {"count": { "$gt": 1 }}}
    ])
)

pprint(pipeline)

[{'_id': {'description': 'Moradia concelho de  Santa Maria da feira com evento '
                         'maio teatro rua imaginárius perto das praias de  '
                         'espinho esmoriz praias Gaia zoo de lourosa e zoo de '
                         'avintes num raio de 10 kms possível de usufruir de '
                         'guia turístico com transporte Cozinha quarto de '
                         'banho terraço exterior Cozinha equipada pátio '
                         'exterior quarto de banho para duches Possibilidade '
                         'de guia para conhecer as zonas circundantes Santa '
                         'Maria da feira praias espinho Gaia Porto Praias '
                         'serra caves de Gaia zoo de avintes parque '
                         'ornotologico Carro mota táxi',
          'name': 'Quarto moradia luxo'},
  'count': 2,
  'documents': ['26556751', '26563602']},
 {'_id': {'description': '',
          'name': 'İstanbul Birden fazla bölüm

There are listings with the same name + description (one case where there is no description):
- Analyze differences to understand if it's the same property.

In [13]:
# Visualize fields that hold different values for listings w/ same name and description

doc1 = db.listingsAndReviews_new.find_one({"_id": "26556751"})
doc2 = db.listingsAndReviews_new.find_one({"_id": "26563602"})

diffs = {}
for key in set(doc1.keys()).union(set(doc2.keys())):  
    if doc1.get(key) != doc2.get(key): 
        diffs[key] = {"doc1": doc1.get(key), "doc2": doc2.get(key)}

pprint(diffs)


{'_id': {'doc1': '26556751', 'doc2': '26563602'},
 'address': {'doc1': {'country': 'Portugal',
                      'country_code': 'PT',
                      'government_area': 'Sanguedo',
                      'location': {'coordinates': [-8.52195, 41.00466],
                                   'is_location_exact': False,
                                   'type': 'Point'},
                      'market': 'Porto',
                      'street': 'Sanguedo, Aveiro, Portugal',
                      'suburb': ''},
             'doc2': {'country': 'Portugal',
                      'country_code': 'PT',
                      'government_area': 'Sanguedo',
                      'location': {'coordinates': [-8.50928, 41.00966],
                                   'is_location_exact': False,
                                   'type': 'Point'},
                      'market': 'Porto',
                      'street': 'Sanguedo, Aveiro, Portugal',
                      'suburb': ''}},
 'ameniti

There are differences in price, however it looks like it's the same property, so we should erase the older one: 26556751.

In [14]:
# Visualize fields that hold different values for listings w/ same name and description

doc1 = db.listingsAndReviews_new.find_one({"_id": "28221594"})
doc2 = db.listingsAndReviews_new.find_one({"_id": "29224880"})
doc3 = db.listingsAndReviews_new.find_one({"_id": "29324861"})

diffs = {}
for key in set(doc1.keys()).union(set(doc2.keys())):  
    if doc1.get(key) != doc2.get(key): 
        diffs[key] = {"doc1": doc1.get(key), "doc2": doc2.get(key), "doc3": doc3.get(key)}

pprint(diffs)


{'_id': {'doc1': '28221594', 'doc2': '29224880', 'doc3': '29324861'},
 'accommodates': {'doc1': 1, 'doc2': 5, 'doc3': 2},
 'address': {'doc1': {'country': 'Turkey',
                      'country_code': 'TR',
                      'government_area': 'Besiktas',
                      'location': {'coordinates': [28.99908, 41.04811],
                                   'is_location_exact': False,
                                   'type': 'Point'},
                      'market': 'Istanbul',
                      'street': 'Beşiktaş, İstanbul, Turkey',
                      'suburb': 'Beşiktaş'},
             'doc2': {'country': 'Turkey',
                      'country_code': 'TR',
                      'government_area': 'Besiktas',
                      'location': {'coordinates': [29.0229, 41.05203],
                                   'is_location_exact': False,
                                   'type': 'Point'},
                      'market': 'Istanbul',
                      'stree

These listings are from different hosts and different property types, so they may be different listings even tho they have the same name.

### Reviews:

In [ ]:
# Reviews with same id for reviews field

pipeline = list(
    db.listingsAndReviews_new.aggregate([
        {"$group": { "_id": "$reviews._id", "count": { "$sum": 1 }} },
        {"$match": {"count": { "$gt": 1 }}},
    ])
)

pprint(pipeline)

[{'_id': [], 'count': 1632}]


- There are no reviews inside the reviews field with the same id.
- The output could represent documents where the reviews._id is missing or documents that don´t have any reviews at all and therefor reviews._id does not exist.


In [ ]:
# Rewiews with some id across reviews, reviews_copy1, reviews_copy2, reviews_copy3 and reviews_copy4

import json

# Find all listings that have the relevant review fields
listings = db.listings.find(
    {"reviews": {"$exists": True}}, 
    {"reviews": 1, "reviews_copy1": 1, "reviews_copy2": 1, "reviews_copy3": 1, "reviews_copy4": 1}
)

different_listings = []

for listing in listings:
    listing_id = listing["_id"]
    
    # Extract sets of review ids 
    review_id_sets = {
        key: set(review["_id"] for review in listing.get(key, []))
        for key in ["reviews", "reviews_copy1", "reviews_copy2", "reviews_copy3", "reviews_copy4"]
        if key in listing
    }

    # Check if all review id sets are identical
    unique_review_sets = set(frozenset(ids) for ids in review_id_sets.values())

    if len(unique_review_sets) > 1:  # Differences 
        different_listings.append(listing_id)

if not different_listings:
    print("All listings have identical review ids across copies.")
else:
    print(f"Total listings with different review ids: {len(different_listings)}")


All listings have identical review IDs across copies.


- We can delete the copies 1, 2, 3 and 4.

## Empty fields

In [16]:

# Check for empty fields in all documents 

cursor = db.listingsAndReviews_new.find()

for doc in cursor:
    empty_fields = {key: doc[key] for key in doc if doc[key] in [None, "", []]}
    
    if empty_fields:
        print(f"Document id: {doc['_id']} empty fields:")
        pprint(empty_fields)


Document id: 10006546 empty fields:
{'host_neighbourhood': ''}
Document id: 10009999 empty fields:
{'access': '',
 'reviews': [],
 'reviews_copy1': [],
 'reviews_copy2': [],
 'reviews_copy3': [],
 'reviews_copy4': []}
Document id: 1001265 empty fields:
{'notes': ''}
Document id: 10021707 empty fields:
{'access': '',
 'host_about': '',
 'house_rules': '',
 'interaction': '',
 'neighborhood_overview': '',
 'notes': '',
 'space': '',
 'transit': ''}
Document id: 10030955 empty fields:
{'access': '',
 'host_about': '',
 'house_rules': '',
 'interaction': '',
 'neighborhood_overview': '',
 'notes': '',
 'reviews': [],
 'reviews_copy1': [],
 'reviews_copy2': [],
 'reviews_copy3': [],
 'reviews_copy4': [],
 'space': '',
 'transit': ''}
Document id: 1003530 empty fields:
{'host_neighbourhood': '', 'interaction': '', 'summary': ''}
Document id: 10038496 empty fields:
{'notes': ''}
Document id: 10047964 empty fields:
{'access': '', 'interaction': '', 'notes': '', 'transit': ''}
Document id: 1005

In [41]:
# Check unique fields that have at least one document where they are empty

empty_field = set()

cursor = db.listingsAndReviews_new.find()

for doc in cursor:
    for key in doc:
        if doc[key] in [None, "", []]: 
            empty_field.add(key)   

pprint(empty_field)


{'access',
 'description',
 'host_about',
 'host_location',
 'host_neighbourhood',
 'house_rules',
 'interaction',
 'name',
 'neighborhood_overview',
 'notes',
 'reviews',
 'reviews_copy1',
 'reviews_copy2',
 'reviews_copy3',
 'reviews_copy4',
 'space',
 'summary',
 'transit'}


- Several documents have empty fields, but these fields are likely not mandatory, we have no further information so we will accept it as it is.
- We can also confirm all listings are associated with a host (because the host field is not in the list), which is expected to be mandatory.

### Nested fields:

In [42]:
# Check for listings that have host_id as empty fields 

cursor = db.listingsAndReviews_new.find({
    "$or": [
        {"host_id": " "},
        {"host_id": None},
        {"host_id": []}
    ]
})

for doc in cursor:
    pprint(doc)

- All hosts have an id.

In [43]:
# Check for listings that have reviews._id as empty fields 

cursor = db.listingsAndReviews_new.find({
    "$or": [
        {"reviews._id": " "},
        {"reviews._id": None},
        {"reviews._id": []}
    ]
})

for doc in cursor:
    pprint(doc)

- All reviews have an id.

In [44]:
# Check for listings that have reviews.reviwer_id as empty fields 

cursor = list(db.listingsAndReviews_new.find({
    "$or": [
        {"reviews.reviwer_id": " "},
        {"reviews.reviwer_id": None},
        {"reviews.reviwer_id": []}
    ]
}))

pprint(len(cursor))

3923


There are 3923 reviews that aren't associated with an user via reviewer_id. This may indicate an inconsistency, but since we have no way to input this information, we will leave it as is.

In [ ]:
# Check for listings that have reviews.listing_id as empty fields 

cursor = list(db.listingsAndReviews_new.find({
    "$or": [
        {"reviews.listing_id": " "},
        {"reviews.listing_id": None},
        {"reviews.listing_id": []}
    ]
}))

pprint(len(cursor))

0


- All reviews are associated with a listing via listing_id.

## Reviews' assignment

In [ ]:
# Check if the listing_id in the reviews matched the document id where the review is nested

pipeline = list(
    db.listingsAndReviews_new.aggregate([
        {"$unwind": "$reviews"},
        {"$match": {"$expr": { "$ne": ["$reviews.listing_id", "$_id"] }}}
    ]))

for doc in pipeline:
    pprint(doc)

- All reviews are inside the correct listing.

## Apply changes to the database

TO DO:

- Create new collection for Reviews with all fields including listing_id. In the collection ListingsAndReviews change name to Listings and keep 3 most recent reviews in the documents of the listings, than a list with the ids of all reviews, e.g. reviews_ids[123233, 33444, (...)]
- Create new collection for transactions w/ listing_id, change buckets to have a bucket for each year.
- Put the host fields in a subset, so that there is a field host and then nested inside that field the host fields.
- Erase this duplicated listing: 26556751.
- Erase reviews_copy1/2/3/4.
- Add an average rating field.
- Indexing.

In [ ]:
listings_collection = db["listingsAndReviews_new"]
reviews_collection = db["Reviews"]
listings_new_collection = db["Listings"]

def migrate_reviews():
    for listing in listings_collection.find():
        listing_id = listing["_id"]
        reviews = listing.get("reviews", [])

        if reviews:
            # Order reviews by date
            reviews_sorted = sorted(reviews, key=lambda x: x["date"], reverse=True)

            for review in reviews:
                review["listing_id"] = listing_id
                reviews_collection.update_one(
                    {"_id": review["_id"]},
                    {"$set": review},
                    upsert=True
                )

            # keep 3 most recent reviews
            Latest_reviews = reviews_sorted[:3]
        else:
            Latest_reviews = []

        listings_new_collection.insert_one({
            "_id": listing["_id"],
            **{k: v for k, v in listing.items() if k != "reviews"},
            "Latest_reviews": Latest_reviews
        })

    db.listingsAndReviews.drop()
    print("Migration completed and old collection dropped")

migrate_reviews()

Migration completed and old collection dropped!


### Check changes

In [ ]:
print("\n Verifying 'Listings' Collection")
pprint(db.Listings.find_one({}, {"_id": 1, "Latest_reviews": 1}))


-- Verifying 'Listings' Collection --
{'_id': '10006546',
 'latest_reviews': [{'_id': '403055315',
                     'comments': 'The house was extremely well located and Ana '
                                 'was able to give us some really great tips '
                                 'on locations to have lunch and eat out. The '
                                 'house was perfectly clean and the easily '
                                 'able to accommodate 6 people despite only '
                                 'having one bathroom. The beds and living '
                                 'room were comfortable. \n'
                                 '\n'
                                 'However, we always felt somewhat on edge in '
                                 'the house due to the number of signs posted '
                                 'around the kitchen, bedrooms and bathroom '
                                 'about being charged 15€ for all sorts of '
              

In [ ]:

print("\nVerifying 'Reviews' Collection")
pprint(db.Reviews.find_one({}))


-- Verifying 'Reviews' Collection --
{'_id': '58663741',
 'comments': 'A casa da Ana e do Gonçalo foram o local escolhido para a '
             'passagem de ano com um grupo de amigos. Fomos super bem '
             'recebidos com uma grande simpatia e predisposição a ajudar com '
             'qualquer coisa que fosse necessário.\r\n'
             'A casa era ainda melhor do que parecia nas fotos, totalmente '
             'equipada, com mantas, aquecedor e tudo o que pudessemos '
             'precisar.\r\n'
             'A localização não podia ser melhor! Não há melhor do que acordar '
             'de manhã e ao virar da esquina estar a ribeira do Porto.',
 'date': datetime.datetime(2016, 1, 3, 5, 0),
 'listing_id': '10006546',
 'reviewer_id': '51483096',
 'reviewer_name': 'Cátia'}


In [ ]:
print("\nVerifying if 'listingsAndReviews' was dropped")
pprint(db.listingsAndReviews_new.find_one({}))


-- Verifying if 'listingsAndReviews' was dropped --
None


In [91]:
duplicate_listing_id = "26556751"

duplicate_listing = listings_new_collection.find_one({"_id": duplicate_listing_id})

if duplicate_listing:
    listings_new_collection.delete_one({"_id": duplicate_listing_id})
    print(f"Listing {duplicate_listing_id} deleted.")

Listing 26556751 deleted.


In [ ]:
reviews_collection.create_index([("listing_id", 1)])
listings_collection.create_index([("price", 1)])


**Patterns used**:

- **Bucket**: Bucket was already being used for the transactions related data, however smaller buckets help limit the size of the data that needs to be processed at once. We believe that it’s more common to analyze transactions by year because properties are usually rented for days/weeks, so we created buckets for each year rather than having a single bucket that contains data across several years. This approach enhances performance, as calculations typically only require the most recent year’s data. 

- **Computed**: Add computed value average_rating for quick retrieval. Since this is a very common value to be retrieved in this type of platform, it is justified that this value is precomputed and saved in the dataset, rather than having to compute it each time that we need to display the value. 

- **Subset**:
We created a Host field which will have a subset with every host related fields that were already present in the database. By structuring the data this way, we can easily access all host-related information, rather than having to search through multiple individual host fields. 

- **Extended reference**:
    - We applied extended reference when creating the Reviews collection, where each review contains the listing_id, and each listing contains an array of review IDs. This will be useful in cases where the user wants to check all reviews for a given listing, or at least more than the ones stored in the listing document. If we only stored listing_id in the reviews, we would need to query the reviews collection to gather all the reviews for a particular listing. However, having the array of review_ids in the listings collection allows to quickly retrieve all reviews for a listing without having to query the reviews collection. 
    - However given that transaction data is not commonly retrieved in most queries, because the main concern is displaying listing information to users, we believe it’s not justified to use extended references in the listing documents for this case.

# 1.1 Database Updates

12) Add a new property with a new host in one of the top 10 cities. The host selects the top 10 most common amenities to list.

13) Add a new review from one of our top 20 reviewers for this new property.

14) Add a new review metric called 'x_factor' with a score of 10. Show that the average score across all metrics is correctly calculated for this listing, using the previously developed query.

In [ ]:
#12
top_cities = list(listings_new_collection.aggregate([
    {"$group": {"_id": "$address.market", "count": {"$sum": 1}}},
    {"$sort": {"count": -1}},
    {"$limit": 10}
]))

chosen_city = top_cities[0]["_id"]
print("Chosen city:", chosen_city)

In [ ]:
top_amenities = list(listings_new_collection.aggregate([
    {"$unwind": "$amenities"},
    {"$group": {"_id": "$amenities", "count": {"$sum": 1}}},
    {"$sort": {"count": -1}},
    {"$limit": 10}
]))

top_amenities = [amenity["_id"] for amenity in top_amenities]
print("Top amenities:", top_amenities)

In [ ]:
new_listing = {
    "listing_url": "https://www.airbnb.com/rooms/new",
    "name": "New Luxury Apartment",
    "summary": f"Brand new luxury apartment in the heart of {chosen_city}",
    "property_type": "Apartment",
    "room_type": "Entire home/apt",
    "bedrooms": 2,
    "beds": 3,
    "bathrooms": 1,
    "accommodates": 4,
    "price": 150.00,
    "amenities": top_amenities,
    "host": {
        "host_id": "12345678",
        "host_name": "New Host",
        "host_since": "2025-03-31"
    },
    "address": {
        "market": chosen_city,
        "country": "CountryName"
    }
}

result = listings_new_collection.insert_one(new_listing)
print("New listing added with _id:", result.inserted_id)


# 2

### Question 1

### Question 2 - Once a month, we reward hosts with recognition. Select three superhosts with at least two listings that can accommodate more than four people.

In [38]:
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$match": { 
            "host.host_is_superhost": True,
            "accommodates": { "$gt": 4 }
        }},
        {"$group": { 
            "_id": "$host.host_id", 
            "host_name": { "$first": "$host.host_name" }, 
            "host_listings_count": { "$sum": 1 }
        }},
        {"$match": { 
            "host_listings_count": { "$gte": 2 } 
        }},
        {"$sort": { "host_listings_count": -1 }}, 
        {"$limit": 3}
    ])
)

print("\nSelected Superhosts:\n")
for doc in pipeline:
    pprint(doc)



Selected Superhosts:

{'_id': '42618840',
 'host_listings_count': 3,
 'host_name': 'Great Vacation Retreats'}
{'_id': '10496350', 'host_listings_count': 3, 'host_name': 'Elite'}
{'_id': '36133410', 'host_listings_count': 3, 'host_name': 'Patty And Beckett'}


### Question 3 - The company considers investing into property to rent. Which bed type is most common in listings with a waterfront and a dishwasher in New York?

In [ ]:
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$match": { 
            "address.market": "New York",
            "amenities": {"$in": ["Dishwasher", "Waterfront"]},
        }},
        {"$group": {
            "_id": "$bed_type", 
            "count": {"$sum": 1}
        }},
        {"$sort": {"count": -1}},
        {"$limit": 1}
    ])
)

for doc in pipeline:
    pprint(doc)


{'_id': 'Real Bed', 'count': 75}


### Question 4 - We're considering hiring someone to write reviews professionally. Who wrote the longest review in New York?

In [54]:
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$match": { 
            "address.market": "New York", 
            "reviews": { "$exists": True } 
        }},
        {"$unwind": "$reviews"},
        {"$sort": { "reviews.comments": -1 }}, 
        {"$project": { 
            "reviews.reviewer_name": 1,
            "reviews.comments": 1
        }},
        {"$limit": 1}
    ])
)

print("\nPerson with the Longest Review in New York:\n")
for doc in pipeline:
    pprint(doc)



Person with the Longest Review in New York:

{'_id': '14569881',
 'reviews': {'comments': 'ﾙｰﾑｵｰﾅｰは非常に日本贔屓のｱﾒﾘｶ人で､部屋はﾃﾞｻﾞｲﾝのｱｲﾃﾞｨｱに富み､清潔で､静かで､ﾘﾗｯｸｽできます｡\n'
                         'ﾙｰﾑﾒｲﾄも皆優しくて親切です｡\n'
                         '泊まった部屋は､大きなｽﾀｼﾞｵの中に手作りで創られた2階建ての小屋の1階でした｡工夫と創造性に溢れた､素敵な空間です｡wifiもあります｡\n'
                         'ただ､もし一般的ﾋﾞｼﾞﾈｽﾕｰｽで考えるなら､この部屋に限らず､一般的なｼｪｱﾙｰﾑは避けたほうがいいでしょう｡日本人はなんだかんだで､仕事に集中したい時は完全に遮蔽されたﾌﾟﾗｲﾍﾞｰﾄ空間を好みます(僕もそう)が､ｼｪｱﾙｰﾑでそれは無理です｡静かとは言え､3人が共同生活をしているので､会話は発生するし､料理や掃除も始まるし､風呂の順番だってﾏﾅｰを守らねばなりません｡あくまでくつろぐときに部屋を利用し､仕事はｶﾌｪなどでやるのが無難です｡\n'
                         'ちなみに､ｵｰﾅｰは日本語を話せますが､少しだけです｡基本的に英語でｺﾐｭﾆｹｰｼｮﾝする必要があります｡\n'
                         'ちなみにｱﾒﾆﾃｨはﾊﾞｽﾀｵﾙ1枚なので､歯ﾌﾞﾗｼや石鹸など身の回りのものは自分で持っていくか近所のﾃﾞﾘなどで買いましょう｡\n'
                         'よい旅を｡',
             'reviewer_name': 'Kazu'}}


### Question 5 - To assess the security of different areas, what is the biggest and smallest (price-security deposit) difference per number of visitors at a property?

In [ ]:
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$match": {
            "price": {"$exists": True},
            "security_deposit": {"$exists": True}  
        }},

        {"$project": {
            "difference": {"$subtract": ["$price", "$security_deposit"]} 
        }},
        {"$group": {
            "_id": None, 
            "max_difference": {"$max": "$difference"}, 
            "min_difference": {"$min": "$difference"}
        }}
    ])
)

print("\nBiggest and Smallest Price-Security Deposit Differences:\n")
for doc in pipeline:
    pprint(doc)




Biggest and Smallest Price-Security Deposit Differences:

{'_id': None,
 'max_difference': Decimal128('8501.00'),
 'min_difference': Decimal128('-38459.00')}


In [ ]:
#mesma coisa mas a retornar os ID's 
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$match": {
            "price": {"$exists": True},  
            "security_deposit": {"$exists": True}  
        }},
        {"$project": {
            "_id": 1,  
            "price": {"$toDouble": "$price"},  
            "security_deposit": {"$toDouble": "$security_deposit"},  
            "difference": {"$subtract": [{"$toDouble": "$price"}, {"$toDouble": "$security_deposit"}]}  
        }},
        {"$sort": {"difference": 1}}, 
        {"$group": {
            "_id": None,  
            "max_difference": {"$last": "$difference"},  
            "max_id": {"$last": "$_id"},
            "min_difference": {"$first": "$difference"},  
            "min_id": {"$first": "$_id"} 
        }}
    ])
)

print("\nBiggest and Smallest Price-Security Deposit Differences:\n")
for doc in pipeline:
    pprint(doc)



Biggest and Smallest Price-Security Deposit Differences:

{'_id': None,
 'max_difference': 8501.0,
 'max_id': '22200454',
 'min_difference': -38459.0,
 'min_id': '2688143'}


In [ ]:
#para ver a db. Depois apagar.
document = db.listingsAndReviews.find({"_id": "2688143"})

print("\nDocument for _id = 22200454:\n")
for doc in document:
    pprint(doc)



Document for _id = 22200454:

{'_id': '2688143',
 'access': '',
 'accommodates': 2,
 'address': {'country': 'Hong Kong',
             'country_code': 'HK',
             'government_area': 'Eastern',
             'location': {'coordinates': [114.19337, 22.28558],
                          'is_location_exact': True,
                          'type': 'Point'},
             'market': 'Hong Kong',
             'street': 'Hong Kong, Hong Kong Island, Hong Kong',
             'suburb': 'Eastern District'},
 'amenities': ['TV',
               'Internet',
               'Wifi',
               'Air conditioning',
               'Kitchen',
               'Paid parking off premises',
               'Elevator',
               'Washer',
               'Dryer',
               'Hangers',
               'Hair dryer',
               'Iron',
               'Laptop friendly workspace',
               'Room-darkening shades',
               'Hot water',
               'Bed linens',
               'Etherne

### Question 6 - Identify areas by whether they are typically used for short breaks, like weekend mini breaks, or whether they are more suitable for long trips. This information support targeted advertising of different customer types. It is not expected to change much over time so we won’t look to update it, we just require current view. What is the average duration of stay (in nights) per type of property per city (you can use the maximum_nights to measure length of stays)? For each property type return the city with the highest and lowest average value.

In [30]:
pipeline = [
    # Ensure maximum_nights exists and is a valid number
    {"$match": {"maximum_nights": {"$exists": True, "$ne": ""}
    }},


    # Group by city and property type, calculating the average maximum_nights
    {"$group": {
        "_id": {"city": "$address.market", "property_type": "$property_type"},
        "avg_maximum_nights": {"$avg": {"$toDouble": "$maximum_nights"}}
    }},

    # Reshape the output for easier processing
    {"$project": {
        "city": "$_id.city",
        "property_type": "$_id.property_type",
        "avg_maximum_nights": 1,
        "_id": 0
    }},

    # Group by property type to find the cities with the highest and lowest average duration
    {"$group": {
        "_id": "$property_type",
        "highest_city": {"$max": {"city": "$city", "avg_nights": "$avg_maximum_nights"}},
        "lowest_city": {"$min": {"city": "$city", "avg_nights": "$avg_maximum_nights"}}
    }}
]

# Execute query
result = list(db.listingsAndReviews.aggregate(pipeline))

# Print result
for item in result:
    pprint(item)


{'_id': 'Boutique hotel',
 'highest_city': {'avg_nights': 577.5, 'city': 'Sydney'},
 'lowest_city': {'avg_nights': 1125.0, 'city': 'Barcelona'}}
{'_id': 'Castle',
 'highest_city': {'avg_nights': 1125.0, 'city': 'Montreal'},
 'lowest_city': {'avg_nights': 1125.0, 'city': 'Montreal'}}
{'_id': 'Pension (South Korea)',
 'highest_city': {'avg_nights': 1125.0, 'city': 'Hong Kong'},
 'lowest_city': {'avg_nights': 1125.0, 'city': 'Hong Kong'}}
{'_id': 'Train',
 'highest_city': {'avg_nights': 1125.0, 'city': 'Sydney'},
 'lowest_city': {'avg_nights': 1125.0, 'city': 'Sydney'}}
{'_id': 'Earth house',
 'highest_city': {'avg_nights': 1125.0, 'city': 'Rio De Janeiro'},
 'lowest_city': {'avg_nights': 1125.0, 'city': 'Rio De Janeiro'}}
{'_id': 'Hut',
 'highest_city': {'avg_nights': 30.0, 'city': 'The Big Island'},
 'lowest_city': {'avg_nights': 30.0, 'city': 'The Big Island'}}
{'_id': 'Condominium',
 'highest_city': {'avg_nights': 505.04761904761904, 'city': 'The Big Island'},
 'lowest_city': {'avg_ni

### Question 7 - We are creating a new webpage for hosts when setting up their account. It will list suggested typical amenities. This data will need to be available every time a host registers a property but is not expected to change very much. The starting point for the list will be all unique amenities currently listed in properties (across all documents). Optimise the database for this use case and show how the data should be queried.

In [16]:
collection = db["listingsAndReviews"]
suggested_amenities = db["suggested_amenities"]

# Extract unique amenities
unique_amenities = collection.aggregate([
    {"$unwind": "$amenities"},
    {"$group": {"_id": "$amenities"}}
])

# Store the unique amenities in a single document
amenities_list = sorted(set(doc["_id"] for doc in unique_amenities))
suggested_amenities.update_one(
    {"_id": "amenities_list"},
    {"$set": {"amenities": amenities_list}},
    upsert=True
)

UpdateResult({'n': 1, 'nModified': 1, 'ok': 1.0, 'updatedExisting': True}, acknowledged=True)

In [18]:
result = db.suggested_amenities.find_one({"_id": "amenities_list"})
print(result["amenities"])

['', '24-hour check-in', 'Accessible-height bed', 'Accessible-height toilet', 'Air conditioning', 'Air purifier', 'Alfresco shower', 'BBQ grill', 'Baby bath', 'Baby monitor', 'Babysitter recommendations', 'Balcony', 'Bath towel', 'Bathroom essentials', 'Bathtub', 'Bathtub with bath chair', 'Beach chairs', 'Beach essentials', 'Beach view', 'Beachfront', 'Bed linens', 'Bedroom comforts', 'Bicycle', 'Bidet', 'Body soap', 'Boogie boards', 'Breakfast', 'Breakfast bar', 'Breakfast table', 'Building staff', 'Buzzer/wireless intercom', 'Cable TV', 'Carbon monoxide detector', 'Cat(s)', 'Ceiling fan', 'Central air conditioning', 'Changing table', "Chef's kitchen", 'Children’s books and toys', 'Children’s dinnerware', 'Cleaning before checkout', 'Coffee maker', 'Convection oven', 'Cooking basics', 'Crib', 'DVD player', 'Day bed', 'Dining area', 'Disabled parking spot', 'Dishes and silverware', 'Dishwasher', 'Dog(s)', 'Doorman', 'Double oven', 'Dryer', 'EV charger', 'Electric profiling bed', 'Elev

### Question 8 - We plan to track our reviewers better. We want to create a webpage that shows the top 20 reviewers and the count of the number of reviews of each of these reviewers. This webpage should be kept up to date. It should also have a link to return the number of reviews for a given reviewer ID or Name (show how to query for number of reviews by ID or query quickly).

In [23]:
reviews_collection = db["listingsAndReviews"]
top_reviewers_collection = db["top_reviewers"]

# ageregate Top 20 Reviewers by Review Count
top_reviewers = reviews_collection.aggregate([
    {"$unwind": "$reviews"}, 
    {"$group": {"_id": "$reviews.reviewer_id", "count": {"$sum": 1}, "reviewer_name": {"$first": "$reviews.reviewer_name"}}},
    {"$sort": {"count": -1}},
    {"$limit": 20} 
])

# store results in `top_reviewers`
top_reviewers_list = [
    {"reviewer_id": doc["_id"], "reviewer_name": doc["reviewer_name"], "review_count": doc["count"]}
    for doc in top_reviewers
]

top_reviewers_collection.delete_many({}) # removes non updates data
top_reviewers_collection.insert_many(top_reviewers_list) # inserts new top 20

print("Top 20 reviewers updated.")

Top 20 reviewers updated.


In [25]:
top_reviewers_list

[{'reviewer_id': '20775242',
  'reviewer_name': 'Filipe',
  'review_count': 24,
  '_id': ObjectId('67ec1ef9354a5acbaa4ce5f6')},
 {'reviewer_id': '67084875',
  'reviewer_name': 'Nick',
  'review_count': 13,
  '_id': ObjectId('67ec1ef9354a5acbaa4ce5f7')},
 {'reviewer_id': '2961855',
  'reviewer_name': 'Uge',
  'review_count': 10,
  '_id': ObjectId('67ec1ef9354a5acbaa4ce5f8')},
 {'reviewer_id': '20991911',
  'reviewer_name': 'Lisa',
  'review_count': 9,
  '_id': ObjectId('67ec1ef9354a5acbaa4ce5f9')},
 {'reviewer_id': '162027327',
  'reviewer_name': 'Thien',
  'review_count': 9,
  '_id': ObjectId('67ec1ef9354a5acbaa4ce5fa')},
 {'reviewer_id': '1705870',
  'reviewer_name': 'David',
  'review_count': 8,
  '_id': ObjectId('67ec1ef9354a5acbaa4ce5fb')},
 {'reviewer_id': '60816198',
  'reviewer_name': 'Todd',
  'review_count': 8,
  '_id': ObjectId('67ec1ef9354a5acbaa4ce5fc')},
 {'reviewer_id': '69140895',
  'reviewer_name': 'Lisa',
  'review_count': 8,
  '_id': ObjectId('67ec1ef9354a5acbaa4ce5fd

In [27]:
reviewer_id = "67084875"  # Example ID
review_count = reviews_collection.count_documents({"reviews.reviewer_id": reviewer_id})
print(f"Reviewer {reviewer_id} has {review_count} reviews.")

Reviewer 67084875 has 1 reviews.


In [37]:
reviewer_name = "Afonso"  # Example name
review_count = reviews_collection.count_documents({"reviews.reviewer_name": reviewer_name})
print(f"Reviewer {reviewer_name} has {review_count} reviews.")

Reviewer Afonso has 1 reviews.


### Question 9 - For each property we store review scores across different metrics (accuracy, check-in, cleanliness etc). We consider adding more metrics, although there is no clarity on what these will be. We want to be able to easily query the average score across all of these metrics, including any new metrics that might be added without changing the query. Adjust the data model so this can be done and show the query for an example property.

In [13]:
pipeline = [
    {"$match": {"review_scores": {"$exists": True, "$ne": {}}}}, #check if review_scores exists and is not null
    {"$project": {"_id": 1, "scores": {"$objectToArray": "$review_scores"}}},
    {"$project": {"_id": 1, "average_score": {"$avg": "$scores.v"}}},
    {"$group": {"_id": "$_id", "average_score": {"$avg": "$average_score"}}}
]



result = list(db.listingsAndReviews.aggregate(pipeline))

for i in result:
    pprint(i)


{'_id': '22574501', 'average_score': 20.0}
{'_id': '16653189', 'average_score': 22.714285714285715}
{'_id': '17613188', 'average_score': 22.857142857142858}
{'_id': '23509750', 'average_score': 22.571428571428573}
{'_id': '31582808', 'average_score': 22.714285714285715}
{'_id': '11567815', 'average_score': 22.571428571428573}
{'_id': '30092212', 'average_score': 22.857142857142858}
{'_id': '25961494', 'average_score': 22.714285714285715}
{'_id': '555588', 'average_score': 22.142857142857142}
{'_id': '6316666', 'average_score': 16.285714285714285}
{'_id': '20580703', 'average_score': 18.571428571428573}
{'_id': '13764019', 'average_score': 22.571428571428573}
{'_id': '3953585', 'average_score': 22.285714285714285}
{'_id': '3288363', 'average_score': 22.857142857142858}
{'_id': '1586454', 'average_score': 18.285714285714285}
{'_id': '24186748', 'average_score': 22.428571428571427}
{'_id': '14462759', 'average_score': 22.857142857142858}
{'_id': '17223695', 'average_score': 22.71428571428

In [14]:
#Example:
pipeline = [
    {"$match": {"_id": "22200454", "review_scores": {"$exists": True, "$ne": []}}},
    {"$unwind": "$review_scores"},
    {"$group": {"_id": "$_id", "average_score": {"$avg": "$review_scores.score"}}}
]

result = list(db.listingsAndReviews.aggregate(pipeline))

print(result)

[{'_id': '22200454', 'average_score': None}]


In [ ]:
#para ver a db. Depois apagar.
document = db.listingsAndReviews.find({"_id": "9505003"})

for doc in document:
    pprint(doc)

{'_id': '9505003',
 'access': 'Your room, the bureau, bathroom, kitchen, veranda and backyard.',
 'accommodates': 2,
 'address': {'country': 'Canada',
             'country_code': 'CA',
             'government_area': 'Rosemont-La Petite-Patrie',
             'location': {'coordinates': [-73.60577, 45.54015],
                          'is_location_exact': True,
                          'type': 'Point'},
             'market': 'Montreal',
             'street': 'Montréal, Québec, Canada',
             'suburb': 'La Petite-Patrie'},
 'amenities': ['Internet',
               'Wifi',
               'Kitchen',
               'Pets live on this property',
               'Dog(s)',
               'Free street parking',
               'Heating',
               'Family/kid friendly',
               'Suitable for events',
               'Washer',
               'Smoke detector',
               'First aid kit',
               'Fire extinguisher',
               'Essentials',
               'Shamp

-> Qual é a escalda das review_scores? porque porque parecem estar todas de 0 a 10 mas a review_cores_rating está de 0 a 100 acho então fazer a average das review_scores n faz sentido

### Question 10 - We aim to have better access to information about transaction, we wish to develop a search engine that can calculate the average value of transactions in a given period of time quickly for a given property.

### Question 11

### Question 12

### Question 13

### Question 14